tutorials/roofit/rf317_global_observables.C

```diff
@@ -0,0 +1,169 @@
+
+   // event observables
+   RooRealVar x("x", "x", -10, 10);
+
+   // parameters
+   RooRealVar mu("mu", "mu", 0.0, -10, 10);
+   RooRealVar sigma("sigma", "sigma", 1.0, 0.1, 2.0);
+
+   // Gaussian model for event observables
+   RooGaussian gauss("gauss", "gauss", x, mu, sigma);
+
+   // global observables (which are no parameters so they are constant)
+   RooRealVar mu_obs("mu_obs", "mu_obs", 1.0, -10, 10);
+   mu_obs.setConstant();
+
+   // constraint pdf
+   RooGaussian constraint("constraint", "constraint", mu_obs, mu, RooConst(0.2));
+
+   // full pdf including constraint pdf
+   RooProdPdf model("model", "model", {gauss, constraint});
+
+   // Generating toy data with randomized global observables
+   // ------------------------------------------------------
+
+   // For most toy-based statistical procedures, it is necessary to also
+   // randomize the global observable when generating toy datasets.
+   //
+   // To that end, let's generate a single event from the model and take the
+   // global observable value (the same is done in the RooStats:ToyMCSampler
+   // class):
+
+   std::unique_ptr<RooDataSet> dataGlob{model.generate({mu_obs}, 1)};
+
+   // Next, we temporarily set the value of `mu_obs` to the randomized value for
+   // generating our toy dataset:
+   double mu_obs_orig_val = mu_obs.getVal();
+
+   RooArgSet{mu_obs}.assign(*dataGlob->get(0));
+
+   // actually generate the toy dataset
+   std::unique_ptr<RooDataSet> data{model.generate({x}, 1000)};
+
+   // When fitting the toy dataset, it is important to set the global
+   // observables in the fit to the values that were used to generate the toy
+   // dataset. To facilitate the bookkeeping of global observable values, you
+   // can attach a snapshot with the current global observable values to the
+   // dataset like this (new feature introduced in ROOT 6.26):
+
+   data->setGlobalObservables({mu_obs});
+
+   // reset original mu_obs value
+   mu_obs.setVal(mu_obs_orig_val);
+
+   // Note that in a future ROOT release, we plan to streamline the sampling of
+   // toy datasets with randomized global observables such that it can be done
+   // with one call to RooAbsPdf::generate():
+   //
+   //   model.generate({x}, 1000, GlobalObservables(mu_obs))
+
+   // Fitting a model with global observables
+   // ---------------------------------------
+
+   // When you fit a model that includes global observables, you need to
+   // specify them in the call to RooAbsPdf::fitTo with the
+   // RooFit::GlobalObservables command argument. By default, the global
+   // observable values attached to the dataset will be prioritized over the
+   // values in the model, so the following fit correctly uses the randomized
+   // global observable values from the toy dataset:
+   std::cout << "1. model.fitTo(*data, GlobalObservables(mu_obs))" << std::endl;
+   std::cout << "------------------------------------------------" << std::endl << std::endl;
```

Comment:
I realised that this usually does a better job than explicit endl, because it will do the right thing based on whether this goes to files or the terminal ...
```suggestion
   std::cout << "1. model.fitTo(*data, GlobalObservables(mu_obs))\n";
   std::cout << "------------------------------------------------\n\n";
```

Reply:
Alright :+1: